src/toil/server/wes/abstract_backend.py

```diff
@@ -0,0 +1,237 @@
+        """
+        List the workflow runs.
+
+        GET /runs
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def run_workflow(self) -> Dict[str, str]:
+        """
+        Run a workflow. This endpoint creates a new workflow run and returns
+        a `RunId` to monitor its progress.
+
+        POST /runs
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_run_log(self, run_id: str) -> Dict[str, Any]:
+        """
+        Get detailed info about a workflow run.
+
+        GET /runs/{run_id}
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def cancel_run(self, run_id: str) -> Dict[str, str]:
+        """
+        Cancel a running workflow.
+
+        POST /runs/{run_id}/cancel
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_run_status(self, run_id: str) -> Dict[str, str]:
+        """
+        Get quick status info about a workflow run, returning a simple result
+        with the overall state of the workflow run.
+
+        GET /runs/{run_id}/status
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def log_for_run(run_id: Optional[str], message: str) -> None:
+        if run_id:
+            logging.info("Workflow %s: %s", run_id, message)
+        else:
+            logging.info(message)
+
+    @staticmethod
+    def secure_path(path: str) -> str:
+        return os.path.join(*[secure_filename(p) for p in path.split("/") if p not in ("", ".", "..")])
+
+    def collect_attachments(self, run_id: Optional[str], temp_dir: Optional[str]) -> Tuple[str, Dict[str, Any]]:
+        """
+        Collect attachments from the current request by staging uploaded files
+        to temp_dir, and return the temp_dir and parsed body of the request.
+
+        :param run_id: The run ID for logging.
+        :param temp_dir: The directory where uploaded files should be staged.
+                         If None, a temporary directory is created.
+        """
+        if not temp_dir:
+            temp_dir = tempfile.mkdtemp()
+        body = {}
+        has_attachments = False
+        for key, ls in connexion.request.files.lists():
+            try:
+                for value in ls:
+                    # uploaded files that are required to execute the workflow
+                    if key == "workflow_attachment":
+                        # guard against maliciously constructed filenames
+                        dest = os.path.join(temp_dir, self.secure_path(value.filename))
+                        if not os.path.isdir(os.path.dirname(dest)):
+                            os.makedirs(os.path.dirname(dest))
+                        self.log_for_run(run_id, f"Staging attachment '{value.filename}' to '{dest}'")
+                        value.save(dest)
+                        has_attachments = True
+                        body[key] = f"file://{temp_dir}"  # Reference to temp working dir.
```

Comment:
Do we maybe have to worry about URL-escaped characters being interpreted by whatever reads this URI, after making it past the safety checks? For example, Firefox is quite happy to open `file:///tmp/%2E%2E/etc/passwd`.

Reply:
Yes and no.. The `secure_filename()` function is smart enough to strip out `%`, but the whole filesystem can be accessed by the workflow code, so we need a way to protect that.

Reply:
I'm not sure that's a problem we can really solve, unless we start containerizing the runner that we launch.